data_service/src/match_processing/match_parser.rs

```diff
@@ -1,31 +1,151 @@
-use riven::models::match_v5::Match;
+use crate::types::{ChampionFrame, MatchTimelineFrame, PuuidToChampionMapping};
+use chrono::{DateTime, TimeDelta};
+use riven::models::match_v5::{
+    Match, MatchTimeline, MatchTimelineInfoFrameParticipantFrame, Participant,
+};
+use std::collections::HashMap;
 
-use crate::types::PuuidToChampionMapping;
+pub async fn parse_match_timeline(
+    match_timeline: MatchTimeline,
+    match_data: Match,
+) -> crate::types::MatchTimeline {
+    let match_frames = match_timeline.info.frames;
 
-pub async fn parse_match(match_data: Match) -> Vec<PuuidToChampionMapping> {
+    // Map of Participant ID to Participant object ( Found in Match Object but missing in MatchTimeline Object)
+    let participant_id_object_map: HashMap<i32, Participant> = match_data
+        .info
+        .participants
+        .iter()
+        .map(|p| (p.participant_id, p.clone()))
+        .collect();
+
+    // Participant ID to Opponent Participant ID Map
+    let participant_id_opponent_id_map: HashMap<i32, i32> = match_data
+        .info
+        .participants
+        .iter()
+        .map(|p| {
+            (
+                p.participant_id,
+                get_opponent_player_id(p.participant_id, match_data.clone()),
+            )
+        })
+        .collect();
+
+    // Using Combined context ( MatchTimeline, Match Objects ) create a new MatchTimeline Object
+    let mapped_frames = match_frames
+        .iter()
+        .map(|frame| {
+            // Weird that participant_frames is a struct of ( x1, x2,...x10) instead of a vector ( something iteratable) so we convert it to a hashmap ( only so that we can use .map())
+            let participants_frame_as_hashmap: HashMap<
+                String,
+                MatchTimelineInfoFrameParticipantFrame,
+            > = serde_json::from_value(
+                serde_json::to_value(frame.participant_frames.clone()).unwrap(),
+            )
+            .unwrap();
+
+            // Mapping Participant Frames to Champion Frames
+            let champion_mappings: Vec<ChampionFrame> = participants_frame_as_hashmap
+                .iter()
+                .map(|(player_id, player)| {
+                    // Current Player and Opposing Player
+                    let current_player = participant_id_object_map
+                        .get(&player_id.parse::<i32>().unwrap())
+                        .unwrap();
+                    let opposing_player = participant_id_object_map
+                        .get(
+                            &participant_id_opponent_id_map
+                                .get(&player_id.parse::<i32>().unwrap())
+                                .unwrap(),
+                        )
+                        .unwrap();
+
+                    // Finally creating the Champion Frame required by the API
+                    ChampionFrame {
```

Comment:
I see you're mapping participant id to champion id but the correct way to do this is probably to call get_puuid_for_champion_mapping and query that data structure instead of duplicating the logic